plugin/manager.go

```diff
@@ -84,6 +82,78 @@ func (m *CLIManager) List(ctx context.Context) ([]string, error) {
 	return plugins, nil
 }
 
+// Install installs a plugin at filePath to the system.
+//
+// It returns the new plugin metadata and a nil eror if and only if
+// the installation succeeded.
+//
+// If plugin already exists, and overwrite is not set, then the new plugin
+// version MUST be higher than the existing plugin version.
+// If overwrite is set, version check is skipped.
+// On sucess, existing plugin metadata and new plugin metadata are returned.
+//
+// If plugin does not exist, directly install the plugin from filePath.
+// On success, the new plugin metadata is returned.
+func (m *CLIManager) Install(ctx context.Context, filePath string, overwrite bool) (*proto.GetMetadataResponse, *proto.GetMetadataResponse, error) {
+	// validate and get new plugin metadata
+	pluginFile := filepath.Base(filePath)
+	pluginName, err := ExtractPluginNameFromFileName(pluginFile)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get plugin name from file path: %w", err)
+	}
+	if err := validatePluginFileExtensionAgainstOS(pluginFile, pluginName); err != nil {
+		return nil, nil, err
+	}
+	newPlugin, err := NewCLIPlugin(ctx, pluginName, filePath)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to create new CLI plugin: %w", err)
+	}
+	newPluginMetadata, err := newPlugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to get metadata of new plugin: %w", err)
+	}
+	// check plugin existence and get existing plugin metadata
+	var existingPluginMetadata *proto.GetMetadataResponse
+	existingPlugin, err := m.Get(ctx, pluginName)
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, nil, fmt.Errorf("failed to check plugin existence: %w", err)
+		}
+	} else { // plugin already exists
+		var err error
+		existingPluginMetadata, err = existingPlugin.GetMetadata(ctx, &proto.GetMetadataRequest{})
+		if err != nil {
+			return nil, nil, fmt.Errorf("failed to get metadata of existing plugin: %w", err)
+		}
```

Comment:
When `overwrite` is true dont call `GetMetadata`. This would allow user to upgrade plugin if somehow plugin file got corrupted.

Reply:
We still need the old plugin's information. What we can do is, when `overwrite` is set, we will continue even `GetMetadata` failed.